Add support for updating the configuration of an existing sync session

tgoyne opened the pull request:
Required for https://github.com/realm/realm-cocoa/pull/6136.

cmelchior commented on src/sync/sync_session.cpp:
```diff
@@ -838,14 +836,34 @@ SyncSession::PublicState SyncSession::state() const
+{
+    while (true) {
+        std::unique_lock<std::mutex> lock(m_state_mutex);
+        if (m_state != &State::inactive) {
```
Why have this condition + loop? Looking at the implementation for `advance_state` it sets the new state immediately? I'm not 100% sure what happens to the `m_session` variable, but if its destroying is delayed, this isn't caught by this anyway.

tgoyne replied:
Advancing to Inactive releases the lock, so after reacquiring the lock we have to recheck that the state did not get switched away from active while we weren't holding the lock.

tgoyne commented:
I added some not-very-good tests for this. I'm not really sure how to test that the new config actually makes it to the sync::Session with the tools we have.

I also fixed some unrelated tests that were sometimes failing when I ran the tests locally.
